### PR TITLE
Fix image upload page for flow with ZIP file containing a multi-component notification

### PR DESCRIPTION
**Pull request by scruti**
[Jira ticket](https://regulatorydelivery.atlassian.net/browse/COSBETA-972)

## Description of first issue
There was a bug where the image upload page in post-brexit multi-component journey gets into an infinite loop requesting image submission and never moves to the next stage of the flow.

This was caused by:
1) Antivirus over the uploaded file running asynchronously in the background.
2) Business logic enforcing the file to have been checked as "safe from virus" in order to load the next page.

Because at submission point the file is still waiting for the AV check, the virus safety enforcement was never accomplished and the user was being redirected back to the image upload page.

### Solution
Allow to load the next page when the antivirus check is pending over the image.

## Description of second issue
Even when the ZIP file contains a multi-component notification, the image upload page when completing the missing information page was displaying information for a single component notification.

 ### Solution
 Display multiple image uploads information (and title) when loading the image upload page for a multi-component notification.
 
 ## Screenshots of new behaviour
 
**Upload images page for multi-components**
image

**Notification details after image submission**
Screenshot from 2020-12-21 16-08-32

**Notification details after multiple image submission**
image

### CODE REVIEW

File: cosmetics-web/app/controllers/responsible_persons/additional_information_controller.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@ def index
       @notification.formulation_file_uploaded!
     end
 
-    if @notification.images_required_and_missing?
+    if @notification.required_images_missing_or_with_virus?
```
Comment (scruti) [inline cosmetics-web/app/controllers/responsible_persons/additional_information_controller.rb:16]: This bit is the actuall bugfix. We allow notifications with images waiting for antivirus check.

File: cosmetics-web/app/controllers/responsible_persons/product_image_upload_controller.rb [hunk 2]
```diff
@@ -4,29 +4,33 @@ class ResponsiblePersons::ProductImageUploadController < SubmitApplicationContro
...
-        render :new
-      end
+    return render_missing_file_error if params[:image_upload].blank?
+
+    params[:image_upload].each { |img| @notification.add_image(img) }
+    if @notification.save
+      redirect_to responsible_person_notification_additional_information_index_path(
+        @notification.responsible_person,
+        @notification,
+      )
```
Comment (scruti) [inline cosmetics-web/app/controllers/responsible_persons/product_image_upload_controller.rb:14]: Changes in this file are just a refactor

File: cosmetics-web/app/controllers/responsible_persons/wizard/notification_build_controller.rb [hunk 3]
```diff
@@ -163,11 +163,7 @@ def render_add_new_component_step
 
   def render_add_product_image_step
     if params[:image_upload].present?
-      params[:image_upload].each do |image|
-        image_upload = @notification.image_uploads.build
-        image_upload.file.attach(image)
-        image_upload.filename = image.original_filename
-      end
+      params[:image_upload].each { |img| @notification.add_image(img) }
```
Comment (scruti) [inline cosmetics-web/app/controllers/responsible_persons/wizard/notification_build_controller.rb:166]: This is just a refactor. Logic moved to `Notification`

File: cosmetics-web/app/models/notification.rb [hunk 4]
```diff
@@ -97,26 +97,31 @@ def reference_number_for_display
...
   end
 
-  def images_are_present_and_safe?
-    !image_uploads.empty? && image_uploads.all?(&:marked_as_safe?)
+  def add_image(image)
+    image_uploads.build.tap do |upload|
+      upload.file.attach(image)
+      upload.filename = image.original_filename
+    end
+  end
```
Comment (scruti) [inline cosmetics-web/app/models/notification.rb:105]: Refactor: Brought from Controller logic

File: cosmetics-web/app/models/notification.rb [hunk 5]
```diff
@@ -97,26 +97,31 @@ def reference_number_for_display
...
+
+  def all_images_passed_anti_virus_check?
+    image_uploads.all?(&:passed_antivirus_check?)
   end
 
   def images_failed_anti_virus_check?
-    image_uploads.any?(&:file_missing?)
+    image_uploads.any?(&:failed_antivirus_check?)
```
Comment (scruti) [inline cosmetics-web/app/models/notification.rb:112]: No idea why the logic for images failing the AV check was relying on the in-existence of the file, as the file still exists after failing antivirus check (it gets flagged as unsafe).

**Comment by nasirkhanbeis** [comment]
Looks good however, for multi-item we need to show image(s) version page which will allow user to upload more than one image. 
image

**Review by nasirkhanbeis — APPROVED** [review]
looking good as expected!

**Review by slawosz — APPROVED** [review]

**Pull request merged**